File: Practical2/C/matrix.c
#include<stdio.h>
#include<stdbool.h>

void populateMatrix(int matrix[5][5], int userInput, int rowTotal, int colTotal);
void print(int matrix[5][5], int rowTotal, int colTotal);
void search(int matrix[5][5], int rowTotal, int colTotal, int item);
int main()
{
	int rowTotal = 5, colTotal = 5, initializeCount = 1;
	
	//create 5x5 array
	int matrix[5][5] = 
	{
		{1,2,3,4,5},
		{6,7,8,9,10},
		{16,17,18,19,20},
		{21,22,23,24,25},
		{26,27,28,29,30}
	};
	
	int userInput = 0;
	printf("Enter number in the range 1 - 10: ");
	scanf("%d", &userInput);
	populateMatrix(matrix, userInput, rowTotal, colTotal);
	printf("Enter item to search for: ");
	scanf("%d", &userInput);
	search(matrix, rowTotal, colTotal, userInput);
	print(matrix, rowTotal, colTotal);
}

void print(int matrix[5][5], int rowTotal, int colTotal)
{
	for(int row = 0; row < rowTotal; ++row)
	{
		for(int col = 0; col < colTotal; ++col)
			printf("%d ", matrix[row][col]);
		
		printf("\n");
	}
}

void search(int matrix[5][5], int rowTotal, int colTotal, int item)
{
	int x, y;
	bool canBreak = false, found = false;
	for(int row = 0; row < rowTotal; ++row)
	{
		for(int col = 0; col < colTotal; ++col)
		{
			if(matrix[row][col] == item)
			{
				x = row;
				y = col;
				canBreak = true;
				found = true;
				break;
			}
		}
		if(canBreak)
			break;
	}
	
	if(found)
	{
		printf("Row: ");
		for(int i = 0; i < colTotal; ++i)
			printf("%d ", matrix[x][i]);
		
		printf("\n");
		printf("Column: ");
		for(int i = 0; i < rowTotal; ++i)
			printf("%d ", matrix[i][y]);
		printf("\n");
	}
	else
		print(matrix, rowTotal, colTotal);
}

void populateMatrix(int matrix[5][5], int userInput, int rowTotal, int colTotal)
{
	int total = 0;
	int diagonalOne = 0, diagonalTwo = 0;
	switch(userInput)
	{
		case 1:
			for(int col = 0; col < colTotal; ++col)
			{
				for(int row = 0; row < rowTotal; ++row)
					total += matrix[row][col];
				
				for(int i = 0; i < colTotal; ++i)
					matrix[i][col] = total;
				
				total = 0;
			}
			break;
		case 2:
			for(int row = 0; row < rowTotal; ++row)
			{
				for(int col = 0; col < colTotal; ++col)
					total += matrix[row][col];
				
				for(int i = 0; i < rowTotal; ++i)
					matrix[row][i] = total;
				
				total = 0;
			}
			break;
		case 4:
			for(int i = 0; i < colTotal; ++i)
			{
				diagonalOne += matrix[i][i];
				diagonalTwo += matrix[i][colTotal-(i+1)];
			}
			
			for(int i = 0; i < colTotal; ++i)
			{
				matrix[i][i] = diagonalOne+diagonalTwo;
				matrix[i][colTotal-(i+1)] = diagonalOne+diagonalTwo;
			}
			break;
		case 8:
			for(int row = 0; row < rowTotal; ++row)
			{
				for(int col = 0; col < colTotal; ++col)
				{
					if(matrix[row][col]%2 == 0)
						matrix[row][col] = matrix[row][col]*matrix[row][col];
					else if(matrix[row][col]%2 != 0)
						matrix[row][col] = matrix[row][col]*matrix[row][col]*matrix[row][col];
				}
			}
	}
}

File: Practical2/gfortran/matrix.f90
program matrix
	implicit none
	Integer :: row
	INTEGER :: col
	integer :: initializeCount
	integer :: userInput
	integer, dimension(5,5) :: myMatrix

	initializeCount = 1
	
	myMatrix = reshape((/1,2,3,4,5,6,7,8,9,10,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30/), shape(myMatrix))
	
	print *, "Enter number in the range 1-10: "
	read *, userInput
	
	if(userInput == 3) then
		userInput = 1
	end if
	
	if(userInput == 5 .or. userInput == 6 .or.  userInput ==7) then
		userInput = 4
	end if
	
	if(userInput == 9 .or. userInput == 10) then
		userInput = 8
	end if
	
	call populateMatrix(myMatrix, userInput)
	
	print *, "Enter item to search for: "
	read *, userInput
	
	call searchMatrix(myMatrix, userInput)
	call printMatrix(myMatrix)
end program matrix

subroutine printMatrix(m)
implicit none
	integer, dimension(5,5) :: m
	integer :: row
	integer :: col
	
	do row = 1, 5
		print *, (m(row,col), col = 1, 5)
	end do
end subroutine printMatrix

subroutine searchMatrix(m, item)
implicit none
	integer, dimension(5,5) :: m
	integer :: row
	integer :: col
	integer :: x
	integer :: y
	integer :: item
	logical :: found
	
	searchLoop: do row = 1, 5
				do col = 1, 5
					if(m(row,col) == item) then
						x = row
						y = col
						found = .true.
						exit searchLoop
					end if
				end do
			end do searchLoop
			
	if(found) then
		print *, "row:"
		print *, (m(x,col), col = 1,5)
		print *, "col:"
		print *, (m(row,y), row = 1,5)
	end if
end subroutine

subroutine populateMatrix(m, userInput)
implicit none
	integer, dimension(5,5) :: m
	integer  :: userInput
	integer  :: total
	integer  :: row
	integer  :: col
	integer  :: i
	integer  :: diagonalOne
	integer  :: diagonalTwo
	
	total = 0
	diagonalOne = 0
	diagonalTwo = 0
	
	select case (userInput)
		case (1)
			do col = 1, 5
				do row = 1, 5
					total = total + m(row,col)
				end do
				do i = 1, 5
					m(i,col) = total
				end do
				total = 0
			end do
		case (2)
			do row = 1, 5
				do col = 1, 5
					total = total + m(row,col)
				end do
				do i = 1, 5
					m(row,i) = total
				end do
				total = 0
			end do
		case (4)
			do i = 1, 5
				diagonalOne = diagonalOne + m(i,i)
				diagonalTwo = diagonalTwo + m(i,6-i)
			end do
			do i = 1, 5
				m(i,i) = diagonalOne + diagonalTwo
				m(i,6-i) = diagonalOne + diagonalTwo
			end do
		case (8)
			do row = 1, 5
				do col = 1, 5
					if(mod(m(row,col), 2) == 0) then
						m(row,col) = m(row,col)*m(row,col)
					else
						m(row,col) = m(row,col)*m(row,col)*m(row,col)
					end if
				end do
			end do
	end select
end subroutine